server/src/main/kotlin/io/provenance/objectstore/gateway/components/EventStreamConsumer.kt

```diff
@@ -1,83 +1,108 @@
+    }
 
-                    val lastProcessedHeight = blockHeightRepository.getLastProcessedBlockHeight()
+    private suspend fun eventStreamLoop() {
+        val netAdapter = okHttpNetAdapter(eventStreamProperties.websocketUri.toString())
+        val lastBlockProcessed = blockHeightRepository.getLastProcessedBlockHeight()
+        blockFlow(netAdapter, decoderAdapter, from = lastBlockProcessed)
+            .collect {  block ->
+                val lastProcessedHeight = blockHeightRepository.getLastProcessedBlockHeight()
 
-                    if (block.txEvents.any { TX_EVENTS.contains(it.eventType) }) { // these are the blocks you are looking for
-                        block.txEvents.forEach {
-                            streamEventHandlerService.handleEvent(it)
-                        }
+                block.blockResult
+                    .txEvents(block.block.header?.dateTime()) { index -> block.block.txData(index) }
+                    .filter { TX_EVENTS.contains(it.eventType) }  // these are the blocks you are looking for
+                    .forEach {
+                        streamEventHandlerService.handleEvent(it)
                     }
 
-                    if (block.height!! < lastProcessedHeight) {
-                        logger.warn("Received lower block height than last processed (${block.height} vs. $lastProcessedHeight)")
-                    } else {
-                        blockHeightRepository.setLastProcessedBlockHeight(block.height!!)
-                    }
+                if (block.height < lastProcessedHeight) {
+                    logger.warn("Received lower block height than last processed (${block.height} vs. $lastProcessedHeight)")
+                } else {
+                    blockHeightRepository.setLastProcessedBlockHeight(block.height)
                 }
             }
+        try {
+            netAdapter.shutdown()
         } catch (e: Exception) {
-            logger.error("Event stream failed", e)
-            eventStreamRunning.set(false)
+            logger.error("Error shutting down netAdatper", e)
+        }
+    }
+
+    private fun tryStartEventStream(
+        eventStreamFn: suspend CoroutineScope.() -> Unit
+    ) {
+        logger.info("EVENTSTREAM INIT")
+        eventStreamScope.launch(Dispatchers.IO) {
+            while (true) {
+                try {
+                    logger.info("EVENTSTREAM START")
+                    eventStreamFn()
+                    logger.info("EVENTSTREAM END/SUCCESS")
+                    eventStreamRunning.set(true)
+                } catch (e: Exception) {
+                    logger.error("EVENTSTREAM END/FAILURE {}", e.message)
+                    eventStreamRunning.set(false)
+                    logger.info("Waiting ${eventStreamProperties.restartDelaySeconds} seconds before reconnecting to event stream")
+                    Thread.sleep(Duration.ofSeconds(eventStreamProperties.restartDelaySeconds).toMillis())
```

Comment:
Nit: Since you're in a suspending anonymous function here, I think you can leverage the coroutine function `delay` here to achieve the same thing but stay "coroutine-ish"

Reply:
oh nice, looks like that will prevent blocking properly maybe, added